overlays/wrap_memory.py

```diff
@@ -0,0 +1,252 @@
+            return ...  # typing this correctly is annoying and not illuminating
+        else:
+            return self.upstream_env.read_only(use_saved_contents_of_dependents)
+
+    @staticmethod
+    def produce_value(key: str, upstream_get: Any, current_env_getter: Any) -> T:
+        "Must be implemented by child environments"
+        raise NotImplementedError()
+
+    def register_dependency(self, key: str, dependency: str) -> None:
+        self.writable_env.dependencies[key] = self.writable_env.dependencies.get(key, set())
+        self.writable_env.dependencies[key].add(dependency)
+
+    def get(self, key: str, dependency: str, use_saved_contents_of_dependents: bool) -> T:
+        self.register_dependency(key, dependency)
+
+        target_cache_table = (self.writable_env.saved_contents_cache_table
+                              if use_saved_contents_of_dependents or module(key) not in self.writable_env.unsaved_modules
+                              else self.writable_env.unsaved_contents_cache_table)
+        # Update the saved_contents_cache_table whether the module is
+        # saved or unsaved.
+        if key not in target_cache_table:
+            target_cache_table[key] = self.produce_value(
+                key,
+                self.upstream_get(use_saved_contents_of_dependents),
+                current_env_getter=target_cache_table.get
+            )
+
+        return target_cache_table[key]
+
+    def update_for_push(self, keys_to_update: Set[str], is_saved_content: bool) -> Set[str]:
+        def update_table(table: Dict[str, T], key: str, use_saved_contents_of_dependents: bool) -> None:
+            table[key] = self.produce_value(
+                key,
+                self.upstream_get(use_saved_contents_of_dependents=use_saved_contents_of_dependents),
+                current_env_getter=table.get
+            )
+
+        downstream_deps = set()
+        for key in keys_to_update:
+            is_unsaved_module = module(key) in self.writable_env.unsaved_modules
+
+            if is_saved_content:
+                # Update saved_contents_cache_table whether the module is saved or unsaved.
+                update_table(self.writable_env.saved_contents_cache_table, key,
+                             use_saved_contents_of_dependents=True)
+
+                # If the module is unsaved, also update
+                # unsaved_contents_cache_table. Newly-saved content needs to
+                # propagate to both saved and unsaved cache tables.
+                if is_unsaved_module:
+                    update_table(self.writable_env.unsaved_contents_cache_table, key,
+                                 use_saved_contents_of_dependents=False)
+            elif is_unsaved_module:
+                update_table(self.writable_env.unsaved_contents_cache_table, key,
+                             use_saved_contents_of_dependents=False)
+
+            downstream_deps |= self.writable_env.dependencies.get(key, set())
```

Comment:
This isn't related to your code at all, but I just realized I forgot to compare before-and-after values; we only update dependencies of the location-invariant-compare shows a change.

I'll need to think at some point about whether that can cause problems. In a pull-based world I'm sure the answer is no but push based is quite a bit harder to do perfectly.

Reply:
Well if we try to do push-based semantics we will need to think pretty hard about what to do if we compute the overlay value for some key and it matches the parent - this is mainly an issue when we first construct the overlay.

Do we even check for this? If we do, then we could theoretically skip triggering dependencies on that first overlay creation. Again, the estimated savings from being smart here are on the order of a half second for that first unsaved change on a reasonably big python file, so maybe not enough to worry about it - we could just do a pull-based cold-start for each overlay.